File: android/app/src/main/kotlin/com/hiddify/hiddify/LocalControlCredentialStore.kt
package com.hiddify.hiddify

import java.io.File
import java.io.RandomAccessFile
import java.security.SecureRandom
import java.util.UUID

data class LocalControlCredential(
    val generation: String,
    val secret: String,
) {
    override fun toString(): String = "LocalControlCredential(generation=$generation, secret=<redacted>)"
}

enum class LocalControlCredentialFailure {
    MISSING,
    CORRUPT,
    UNAVAILABLE,
}

class LocalControlCredentialException(
    val failure: LocalControlCredentialFailure,
    cause: Throwable? = null,
) : Exception("Protected control credential is unavailable", cause)

interface LocalControlCredentialPersistence {
    fun load(): ByteArray?
    fun insertIfAbsent(candidate: ByteArray): ByteArray
}

object LocalControlFileLock {
    private val processMonitor = Any()

    fun <T> withLock(lockFile: File, action: () -> T): T = synchronized(processMonitor) {
        lockFile.parentFile?.mkdirs()
        RandomAccessFile(lockFile, "rw").channel.use { channel ->
            channel.lock().use { action() }
        }
    }
}

class LocalControlCredentialStore(
    private val persistence: LocalControlCredentialPersistence,
    private val randomBytes: (Int) -> ByteArray = ::secureRandomBytes,
    private val generation: () -> String = { UUID.randomUUID().toString() },
) {
    fun loadExisting(): LocalControlCredential {
        val stored = persistence.load()
            ?: throw LocalControlCredentialException(LocalControlCredentialFailure.MISSING)
        return decode(stored)
    }

    fun loadOrCreate(): LocalControlCredential {
        persistence.load()?.let { return decode(it) }
        val credential = LocalControlCredential(
            generation = generation(),
            secret = randomBytes(32).joinToString("") { "%02x".format(it.toInt() and 0xff) },
        )
        if (!credential.isValid()) {
            throw LocalControlCredentialException(LocalControlCredentialFailure.UNAVAILABLE)
        }
        return decode(persistence.insertIfAbsent(encode(credential)))
    }

    private fun encode(credential: LocalControlCredential): ByteArray =
        "1\n${credential.generation}\n${credential.secret}".toByteArray(Charsets.UTF_8)

    private fun decode(stored: ByteArray): LocalControlCredential {
        val parts = stored.toString(Charsets.UTF_8).split('\n')
        if (parts.size != 3 || parts[0] != "1") {
            throw LocalControlCredentialException(LocalControlCredentialFailure.CORRUPT)
        }
        val credential = LocalControlCredential(generation = parts[1], secret = parts[2])
        if (!credential.isValid()) {
            throw LocalControlCredentialException(LocalControlCredentialFailure.CORRUPT)
        }
        return credential
    }

    private fun LocalControlCredential.isValid(): Boolean =
        generation.isNotEmpty() &&
            generation.length <= 128 &&
            '\n' !in generation &&
            secret.length == 64 &&
            secret.all { it in '0'..'9' || it in 'a'..'f' }

    companion object {
        private val secureRandom = SecureRandom()

        private fun secureRandomBytes(count: Int) = ByteArray(count).also(secureRandom::nextBytes)
    }
}

File: ios/Shared/LocalControlCredentialStore.swift
import Foundation
import Security

struct LocalControlCredential: Codable, Equatable {
    let version: Int
    let generation: String
    let secret: String

    init(generation: String, secret: String) {
        self.version = 1
        self.generation = generation
        self.secret = secret
    }

    var isValid: Bool {
        version == 1 &&
            !generation.isEmpty &&
            secret.count == 64 &&
            secret.unicodeScalars.allSatisfy { CharacterSet(charactersIn: "0123456789abcdef").contains($0) }
    }
}

enum LocalControlCredentialError: Error {
    case missing
    case corrupt
    case unavailable

    var nsError: NSError {
        let code: Int
        switch self {
        case .missing: code = 1
        case .corrupt: code = 2
        case .unavailable: code = 3
        }
        return NSError(
            domain: "LocalControlCredential",
            code: code,
            userInfo: [NSLocalizedDescriptionKey: "Protected control credential is unavailable"]
        )
    }
}

protocol LocalControlCredentialPersistence {
    func load() throws -> Data?
    func insertIfAbsent(_ candidate: Data) throws -> Data
}

struct LocalControlCredentialStore {
    static let shared = LocalControlCredentialStore(
        persistence: KeychainLocalControlCredentialPersistence(
            accessGroup: FilePath.groupName
        )
    )

    private let persistence: LocalControlCredentialPersistence
    private let randomBytes: (Int) throws -> Data
    private let generation: () -> String

    init(
        persistence: LocalControlCredentialPersistence,
        randomBytes: @escaping (Int) throws -> Data = LocalControlCredentialStore.secureRandomBytes,
        generation: @escaping () -> String = { UUID().uuidString.lowercased() }
    ) {
        self.persistence = persistence
        self.randomBytes = randomBytes
        self.generation = generation
    }

    func loadExisting() throws -> LocalControlCredential {
        guard let data = try persistence.load() else {
            throw LocalControlCredentialError.missing
        }
        return try decode(data)
    }

    func loadOrCreate() throws -> LocalControlCredential {
        if let data = try persistence.load() {
            return try decode(data)
        }
        let credential = LocalControlCredential(
            generation: generation(),
            secret: try randomBytes(32).map { String(format: "%02x", $0) }.joined()
        )
        guard credential.isValid else {
            throw LocalControlCredentialError.unavailable
        }
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        let stored = try persistence.insertIfAbsent(encoder.encode(credential))
        return try decode(stored)
    }

    private func decode(_ data: Data) throws -> LocalControlCredential {
        guard
            let credential = try? JSONDecoder().decode(LocalControlCredential.self, from: data),
            credential.isValid
        else {
            throw LocalControlCredentialError.corrupt
        }
        return credential
    }

    private static func secureRandomBytes(count: Int) throws -> Data {
        var bytes = [UInt8](repeating: 0, count: count)
        guard SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes) == errSecSuccess else {
            throw LocalControlCredentialError.unavailable
        }
        return Data(bytes)
    }
}

private struct KeychainLocalControlCredentialPersistence: LocalControlCredentialPersistence {
    private let service = "com.womaninred.local-control"
    private let account = "active-session-v1"
    private let accessGroup: String

    init(accessGroup: String) {
        self.accessGroup = accessGroup
    }

    func load() throws -> Data? {
        var query = baseQuery
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne
        var result: CFTypeRef?
        switch SecItemCopyMatching(query as CFDictionary, &result) {
        case errSecSuccess:
            guard let data = result as? Data else {
                throw LocalControlCredentialError.corrupt
            }
            return data
        case errSecItemNotFound:
            return nil
        default:
            throw LocalControlCredentialError.unavailable
        }
    }

    func insertIfAbsent(_ candidate: Data) throws -> Data {
        var query = baseQuery
        query[kSecValueData as String] = candidate
        query[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        switch SecItemAdd(query as CFDictionary, nil) {
        case errSecSuccess:
            return candidate
        case errSecDuplicateItem:
            guard let existing = try load() else {
                throw LocalControlCredentialError.unavailable
            }
            return existing
        default:
            throw LocalControlCredentialError.unavailable
        }
    }

    private var baseQuery: [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecAttrAccessGroup as String: accessGroup,
            kSecAttrSynchronizable as String: false,
        ]
    }
}
